Several small fixes

dNechita opened the pull request:
## PR Description

Several small fixes came up along the way and added up to a noticeable change.

## PR Type
- [x] Bug fix (a change that fixes an issue)
- [ ] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have commented new code, particularly complex or unclear areas
- [ ] I have checked that I did not introduce new warnings or errors (CI output)
- [ ] I have checked that components that use libiio did not get broken
- [ ] I have updated the documentation accordingly (GitHub Pages, READMEs, etc)

dNechita commented on README_BUILD.md:
```diff
@@ -48,7 +48,7 @@ Cmake Options          | Default | Target | Description
 `PYTHON_BINDINGS`      | OFF |        All | Install PYTHON bindings                            |
 `WITH_UTILS`           |  ON |        All | Build the utility programs (iio-utils)           |
-`WITH_EXAMPLES`        | OFF |        All | Build the example programs                         |
+`WITH_EXAMPLES`        | OFF |        Linux, MaC | Build the example programs                         |
```
Actually, they work on Windows, but one example needs a bit of tweaking. I will remove this change.
Thanks!

dNechita commented:
V1:

- Undo the statement that examples work only on Linux and Mac and not Windows
- Fix adrv9002 not building on windows
- Update the list of examples in examples/README by adding the adrv9002-iiostream
- Update examples: adrv9002-iiostream, adrv9009-iiostream and ad9371-iiostream to take an URI via command line

nunojsa commented on README_BUILD.md:
```diff
@@ -102,7 +102,7 @@ Cmake Options     | Default | Description                                    |
 |                 |         | Debug   : Print debug/info/warnings/errors (very verbose)  |
 
-Options which effect iiod only. These are only available on Linux.
+Options which affect iiod only. These are only available on Linux.
```
Oh, not my point :). The `:` separators is for kind of directory walk. So having `README_BUILD.md: Fix grammar: ` is odd. What I meant was something like `README_BUILD.md: Fix grammar (use  'affect' instead of 'effect')`. I would also strip the files suffix from the git subject. So, just `README_BUILD: ...`

nunojsa commented on examples/adrv9002-iiostream.c:
```diff
@@ -18,13 +18,11 @@
-#define info(...) \
-	printf("%s, %d: INFO: " FMT(__VA_ARGS__, 0)"%s", __func__, __LINE__, ARGS(__VA_ARGS__, ""))
+#define info(fmt, ...) \
+	printf("%s, %d: INFO: " fmt, __func__, __LINE__, ##__VA_ARGS__)
```
Be careful with these changes... Did you tested it on some older windows compilers? IIRC, I had to do all of that gymnastics with the macros exactly because of windows.

AFAIK, ##__VA_ARGS__ is a gnu extension and I thought we were not using it? So I'm a bit confused how the above compiles on window. With what compiler did you tested it? Maybe newer compilers support it...

dNechita replied:
Is there a reason for why we need the examples to be built without GNU extensions?

nunojsa replied:
Note that I'm aware. I guess it's just a matter of portability since the examples should run in multiple platforms.

> The ##_VA_ARGS is a GNU extension indeed but has been standardized in C2x.

Yes, I assumed you're compiling using a recent standard.

> I wouldn't worry about this not working with older windows compilers, as long as it's not part of the library itself.

I'm not going to argue much here as I'm not aware about what is our policy (commitment) in terms of compilers (and C standards) support. But this always opens the door to get someone opening issues or complaining a given piece of code does not work in the environment.

nunojsa commented on examples/adrv9009-iiostream.c:
```diff
@@ -206,9 +213,16 @@ int main (__notused int argc, __notused char **argv)
+		err = iio_err(ctx);
+		IIO_ENSURE(!err && "No context");
+	}
+	else if (argc == 2) {
```
} else if (...)?

nunojsa commented on examples/adrv9002-iiostream.c:
```diff
@@ -226,11 +233,22 @@ int main(void)
+			return EXIT_FAILURE;
+		}
+	}
+	else if (argc == 2) {
```
ditto

nunojsa commented on examples/ad9371-iiostream.c:
```diff
@@ -188,7 +188,14 @@ bool cfg_ad9371_streaming_ch(struct stream_cfg *cfg, enum iodev type, int chid)
+ * URI context, find out the uri by typing `iio_info -s` at the command line of the host PC
+ $./a.out usb:x.x.x
+ */
+int main (int argc, char **argv)
```
Not sure if we should not have this a patch per file...

nunojsa reviewed (commented):
My main concern is with the macro change in the adrv9002 iio stream example.

nunojsa commented:
> It would make sure the examples stay buildable in the future. I will update CI so that at least one job builds them.

I would say that we should enable them at least:

 * one Linux distro;
 * all macOs jobs;
 * all windows jobs;

On linux we can be a bit more confident that if the examples build on one distro, they should be build in the others as well. For the other OSs I just don't know them enough so I would be careful. OTHO, building the examples should be fast enough so we could just enabled it everywhere.

nunojsa commented:
> I enabled the examples on all windows jobs where a few warnings are being thrown making the jobs fail. I need to figure out whether to fix them or not treat them as errors.

yes, I suspected that would happen

dNechita commented:
V2:

- Removed ##__VA_ARGS as it depends on GNU extensions when using older compilers (that don't fully use the C23 standard)
- Fixed warnings (that are treated as errors) about code `return 0;` that never gets reached
- Fixed warnings about implicit conversions between `ssize_t` and `int`
- Changed cmake such that wingetopt gets compiled into a separate library and each example link to it. This helps control how wingetopt gets compiled allowing flexibility when needing to silence warnings, etc.
- The `shutdown()` that is used by many examples will now exit with code 1 (instead of 0) to indicate an error occurred.
- Enabled the build of examples on CI, for Ubuntu 24.04 and for all MSVC builds. Observed that examples were already enabled for all MacOS builds and for all ARM builds.

nunojsa reviewed (commented):
LGTM. Only complain is with the lack of consistency with git subjects. Like we have:

`examples: ...`
`example/ad9361: ...`

And If I'm not mistaken I think the most used style in libiio is similar to linux. So it would be eg: `example: ad9361-iiostream: ...`. 

Another mismatch is `cmake: ...`  vs `CMake:...`. I would just go with the most used style.

I know the above might be a pain in the ass given the amount of patches in the PR and that I might be annoying but I do really believe that consistency matters in your git log.

dNechita commented:
> LGTM. Only complain is with the lack of consistency with git subjects. Like we have:
> 
> `examples: ...` `example/ad9361: ...`
> 
> And If I'm not mistaken I think the most used style in libiio is similar to linux. So it would be eg: `example: ad9361-iiostream: ...`.
> 
> Another mismatch is `cmake: ...` vs `CMake:...`. I would just go with the most used style.
> 
> I know the above might be a pain in the ass given the amount of patches in the PR and that I might be annoying but I do really believe that consistency matters in your git log.

I appreciate the feedback. The subjects of some of my commits could use a bit more consistency.
Fixed!
